### PR TITLE
Add combiner_run

### PR DESCRIPTION
**Pull request by yashykt**
Add combiner_run to run closures immediately if we already have the combiner lock.

This PR also rectifies cases where we were using GRPC_CLOSURE_RUN for a closure to be run on a combiner when we don't already have the lock on the combiner.

### CODE REVIEW

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE                                      FILE SIZE
 ++++++++++++++ GROWING                        ++++++++++++++
  +0.0%     +88 [None]                            +352  +0.0%
  +1.8%     +48 src/core/lib/iomgr/combiner.cc     +48  +1.8%
       +23%     +23 [Unmapped]                         +23   +23%
      [NEW]     +23 combiner_finally_run               +23  [NEW]
      [NEW]      +2 combiner_run                        +2  [NEW]

  +0.0%    +136 TOTAL                             +400  +0.0%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

File: src/core/lib/iomgr/combiner.cc [hunk 1]
```diff
@@ -343,6 +345,39 @@ static void combiner_finally_exec(grpc_closure* closure, grpc_error* error) {
...
+static void combiner_run(grpc_closure* closure, grpc_error* error) {
+#ifndef NDEBUG
+  closure->scheduled = false;
+  grpc_combiner* lock = COMBINER_FROM_CLOSURE_SCHEDULER(closure, scheduler);
+  GRPC_COMBINER_TRACE(gpr_log(
+      GPR_DEBUG,
+      "Combiner:%p grpc_combiner_run closure:%p created [%s:%d] run [%s:%d]",
+      lock, closure, closure->file_created, closure->line_created,
+      closure->file_initiated, closure->line_initiated));
+  GPR_ASSERT(grpc_core::ExecCtx::Get()->combiner_data()->active_combiner ==
```
[the remark this answers is not in the document]
Comment (yashykt) [reply src/core/lib/iomgr/combiner.cc]: grpc_core::ExecCtx::Get() is a pointer thread local so there are no atomics there. As for being blind to bugs, I was hoping that all such scheduling issues (which are going to be programmer errors) would be caught in our tests in debug mode.

Also, it won't be correct to run a closure of a different combiner/scheduler while running inside a combiner. That would lead to all kinds of synchronization issues. A better option would be to just use closure_sched in that case. This also means that we can't just blindly convert closure_sched to closure_run everywhere.

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE                                      FILE SIZE
 ++++++++++++++ GROWING                        ++++++++++++++
  +0.0%     +40 [None]                            +224  +0.0%
  +1.2%     +32 src/core/lib/iomgr/combiner.cc     +32  +1.2%
      [NEW]     +23 combiner_run                       +23  [NEW]
      +9.1%      +9 [Unmapped]                          +9  +9.1%

  +0.0%     +72 TOTAL                             +256  +0.0%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE                                      FILE SIZE
 ++++++++++++++ GROWING                        ++++++++++++++
  +0.0%    +152 [None]                            +504  +0.0%
  +4.8%    +128 src/core/lib/iomgr/combiner.cc    +128  +4.8%
      [NEW]    +115 combiner_run                      +115  [NEW]
       +13%     +13 [Unmapped]                         +13   +13%

  +0.0%    +280 TOTAL                             +632  +0.0%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE                                      FILE SIZE
 ++++++++++++++ GROWING                        ++++++++++++++
  +0.0%    +152 [None]                            +448  +0.0%
  +4.8%    +128 src/core/lib/iomgr/combiner.cc    +128  +4.8%
      [NEW]    +115 combiner_run                      +115  [NEW]
       +13%     +13 [Unmapped]                         +13   +13%

  +0.0%    +280 TOTAL                             +576  +0.0%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

File: src/core/lib/iomgr/combiner.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (yashykt) [reply src/core/lib/iomgr/combiner.cc]: I could also let combiner_run fallback on combiner_exec if the condition is not satisfied instead of asserting out.

File: src/core/ext/transport/chttp2/transport/chttp2_transport.cc [hunk 2]
```diff
@@ -1665,8 +1665,8 @@ static void cancel_pings(grpc_chttp2_transport* t, grpc_error* error) {
 static void send_ping_locked(grpc_chttp2_transport* t,
                              grpc_closure* on_initiate, grpc_closure* on_ack) {
   if (t->closed_with_error != GRPC_ERROR_NONE) {
-    GRPC_CLOSURE_SCHED(on_initiate, GRPC_ERROR_REF(t->closed_with_error));
-    GRPC_CLOSURE_SCHED(on_ack, GRPC_ERROR_REF(t->closed_with_error));
+    GRPC_CLOSURE_RUN(on_initiate, GRPC_ERROR_REF(t->closed_with_error));
```
Comment (markdroth) [inline src/core/ext/transport/chttp2/transport/chttp2_transport.cc]: Isn't this a case where we are currently holding the combiner but are invoking closures that do not necessarily run in the combiner?  Given that, we probably want to use `GRPC_CLOSURE_SCHED()` here.

File: src/core/lib/iomgr/combiner.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (markdroth) [reply src/core/lib/iomgr/combiner.cc]: I think that as long as we have separate `GRPC_CLOSURE_RUN()` and `GRPC_CLOSURE_SCHED()` calls and require callers to choose the right one, then we should make this an assertion to ensure that all callers are doing the right thing.  However, if we ever decided to switch to a model where we combine the two calls and expect the implementation to do the right thing based on the scheduler, then we could have the implementation switch dynamically.

File: src/core/ext/transport/chttp2/transport/chttp2_transport.cc (hunk 2 above)
Comment (yashykt) [reply src/core/ext/transport/chttp2/transport/chttp2_transport.cc]: Oh yeah, this was like a canary to check that the new combiner run still works. (I originally also thought that this would be a permanent change - I didn't take into consideration channel pings.) I'll revert this.

**Comment by yashykt** [comment]
As for benchmarks, I have not run the full benchmark suite, but by running a small microbenchmark around GRPC_CLOSURE_SCHED and GRPC_CLOSURE_RUN, I saw that combiner_run saves 90% off of combiner_exec/

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE                                      FILE SIZE
 ++++++++++++++ GROWING                        ++++++++++++++
  +0.0%    +152 [None]                            +456  +0.0%
  +4.8%    +128 src/core/lib/iomgr/combiner.cc    +128  +4.8%
      [NEW]    +115 combiner_run                      +115  [NEW]
       +13%     +13 [Unmapped]                         +13   +13%

  +0.0%    +280 TOTAL                             +584  +0.0%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE                                      FILE SIZE
 ++++++++++++++ GROWING                        ++++++++++++++
  +0.0%    +152 [None]                            +456  +0.0%
  +4.8%    +128 src/core/lib/iomgr/combiner.cc    +128  +4.8%
      [NEW]    +115 combiner_run                      +115  [NEW]
       +13%     +13 [Unmapped]                         +13   +13%

  +0.0%    +280 TOTAL                             +584  +0.0%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

File: src/core/ext/filters/client_channel/client_channel.cc [hunk 3]
```diff
@@ -3242,7 +3242,7 @@ static void on_external_watch_complete_locked(void* arg, grpc_error* error) {
                            "external_connectivity_watcher");
   external_connectivity_watcher_list_remove(w->chand, w);
   gpr_free(w);
-  GRPC_CLOSURE_RUN(follow_up, GRPC_ERROR_REF(error));
+  GRPC_CLOSURE_SCHED(follow_up, GRPC_ERROR_REF(error));
```
Comment (markdroth) [inline src/core/ext/filters/client_channel/client_channel.cc]: Why is this being changed?  It looks to me like both the caller and the closure are running in the combiner, so this is the kind of place we want to use run.

Comment (yashykt) [reply src/core/ext/filters/client_channel/client_channel.cc]: https://github.com/grpc/grpc/blob/fec4b30d9a94cb6700bfe92ab50a3eccb9ebf128/src/core/ext/filters/client_channel/channel_connectivity.cc#L221

is one place where the closure being run is not initialized for the combiner. 
(This change was actually motivated because of a test failure which led to debugging and changing this.)

File: src/core/lib/iomgr/combiner.cc [hunk 4]
```diff
@@ -63,11 +63,12 @@ struct grpc_combiner {
...
 
+static void combiner_run(grpc_closure* closure, grpc_error* error);
 static void combiner_exec(grpc_closure* closure, grpc_error* error);
 static void combiner_finally_exec(grpc_closure* closure, grpc_error* error);
 
 static const grpc_closure_scheduler_vtable scheduler = {
-    combiner_exec, combiner_exec, "combiner:immediately"};
+    combiner_run, combiner_exec, "combiner:immediately"};
 static const grpc_closure_scheduler_vtable finally_scheduler = {
     combiner_finally_exec, combiner_finally_exec, "combiner:finally"};
```
Comment (markdroth) [inline src/core/lib/iomgr/combiner.cc]: Do we want to make the same change for the finally scheduler?

File: src/core/lib/surface/call.cc [hunk 5]
```diff
@@ -1259,8 +1259,8 @@ static void post_batch_completion(batch_control* bctl) {
   if (bctl->completion_data.notify_tag.is_closure) {
     /* unrefs bctl->error */
     bctl->call = nullptr;
-    GRPC_CLOSURE_RUN((grpc_closure*)bctl->completion_data.notify_tag.tag,
-                     error);
+    GRPC_CLOSURE_SCHED((grpc_closure*)bctl->completion_data.notify_tag.tag,
```
Comment (markdroth) [inline src/core/lib/surface/call.cc]: This change seems unrelated to the rest of this PR, since there's no combiner here.  What's the reason for this?

File: src/core/lib/iomgr/combiner.cc (hunk 4 above)
Comment (yashykt) [reply src/core/lib/iomgr/combiner.cc]: The finally scheduler is a different beast. It is going to require a different thinking and a different set of rules.

File: src/core/lib/surface/call.cc (hunk 5 above)
Comment (yashykt) [reply src/core/lib/surface/call.cc]: There are usecases where this closure is to be run inside in a combiner. As you point out, we are not in a combiner here, so this needed to be changed for correctness.

Comment (markdroth) [reply src/core/lib/surface/call.cc]: Oh right -- this happens from the grpclb code.  I forgot about that.

Might be worth adding a comment about why this uses sched instead of run.

Comment (yashykt) [reply src/core/lib/surface/call.cc]: Done!

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by yashykt** [comment]
Known issues : #14182 #14966 #14730 #13611

**Comment by yashykt** [comment]
Thanks for reviewing!
